File: plugins/ticket/skills/ticket/scripts/providers/github.js
// github provider — file the Canonical Ticket Model as a GitHub issue via the `gh` CLI.
// Config: an authenticated `gh` CLI (`gh auth status`). No secrets read from disk.
// Options: dryRun, repo (OWNER/NAME), noLabels.
'use strict';
const { spawnSync } = require('child_process');
const { renderMarkdown } = require('../lib/render.js');

function labelsFor(t) {
  return [`type:${t.type || 'task'}`, `priority:${t.priority || 'med'}`, ...(t.labels || [])];
}

function gh(args, input) {
  return spawnSync('gh', args, { input, encoding: 'utf8', windowsHide: true });
}

// Dedupe via the plain list API, not --search: GitHub's search index lags behind
// issue creation by seconds to minutes, so a just-filed duplicate would slip through.
// The list endpoint is real-time; exact-title matching happens client-side.
function listIssues(repo) {
  const args = ['issue', 'list', '--state', 'open', '--limit', '100',
                '--json', 'number,title,url'];
  if (repo) args.push('--repo', repo);
  const r = gh(args);
  if (r.error || r.status !== 0) return [];
  try { return JSON.parse(r.stdout || '[]'); } catch { return []; }
}

async function emit(ticket, opts) {
  const title = (ticket.title || '').trim();
  if (!title) throw new Error('ticket has no title');

  const body = renderMarkdown(ticket, { includeTitle: false });
  const labels = opts.noLabels ? [] : labelsFor(ticket);

  if (opts.dryRun) {
    const preview = [
      '[github dry-run] gh issue create',
      `  repo:     ${opts.repo || '<current repo>'}`,
      `  title:    ${title}`,
      `  labels:   ${labelsFor(ticket).join(' ')}`,
      `  assignee: ${ticket.assignee || '<none>'}`,
      '--- body ---',
      body,
    ];
    return { preview: preview.join('\n') };
  }

  const probe = gh(['--version']);
  if (probe.error) throw new Error('gh CLI not found on PATH');

  const dup = listIssues(opts.repo).find((i) => i.title === title);
  if (dup) {
    process.stderr.write(
      `github: an open issue with this title already exists: ${dup.url} (skipping create)\n`);
    return { provider: 'github', ref: String(dup.number), url: dup.url, deduped: true };
  }

  const create = (withLabels) => {
    const args = ['issue', 'create'];
    if (opts.repo) args.push('--repo', opts.repo);
    args.push('--title', title, '--body-file', '-');
    if (withLabels) for (const l of labels) args.push('--label', l);
    if (ticket.assignee) args.push('--assignee', ticket.assignee);
    return gh(args, body);
  };

  let r = create(labels.length > 0);
  if (r.status !== 0 && labels.length && /label/i.test(r.stderr || '')) {
    process.stderr.write(
      'github: some labels do not exist in the repo; retrying without labels ' +
      '(type/priority stay in the body)\n');
    r = create(false);
  }
  if (r.error) throw new Error(`gh failed to run: ${r.error.message}`);
  if (r.status !== 0) throw new Error((r.stderr || '').trim() || 'gh issue create failed');

  const lines = r.stdout.trim().split('\n');
  const url = lines[lines.length - 1].trim();
  const m = url.match(/(\d+)\s*$/);
  return { provider: 'github', ref: m ? m[1] : '', url };
}

module.exports = { emit };

File: plugins/ticket/skills/ticket/scripts/providers/github.py
#!/usr/bin/env python3
"""github provider — file the Canonical Ticket Model as a GitHub issue via the `gh` CLI.

Config: an authenticated `gh` CLI (`gh auth status`). No secrets read from disk.
Options: dry_run, repo (OWNER/NAME), no_labels.
"""
import json
import pathlib
import re
import shutil
import subprocess
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent.parent))
from lib.render import render_markdown  # noqa: E402


def _labels(ticket: dict) -> list[str]:
    return [f"type:{ticket.get('type', 'task')}",
            f"priority:{ticket.get('priority', 'med')}"] + list(ticket.get("labels") or [])


def _list_issues(title: str, repo: str | None) -> list[dict]:
    cmd = ["gh", "issue", "list", "--search", f"{title} in:title", "--state", "open",
           "--limit", "10", "--json", "number,title,url"]
    if repo:
        cmd += ["--repo", repo]
    r = subprocess.run(cmd, capture_output=True, text=True)
    if r.returncode != 0:
        return []
    try:
        return json.loads(r.stdout or "[]")
    except json.JSONDecodeError:
        return []


def emit(ticket: dict, opts: dict) -> dict:
    if not shutil.which("gh"):
        raise RuntimeError("gh CLI not found on PATH")
    title = (ticket.get("title") or "").strip()
    if not title:
        raise RuntimeError("ticket has no title")

    body = render_markdown(ticket, include_title=False)
    repo = opts.get("repo")
    labels = [] if opts.get("no_labels") else _labels(ticket)
    assignee = ticket.get("assignee")

    if opts.get("dry_run"):
        preview = [
            "[github dry-run] gh issue create",
            f"  repo:     {repo or '<current repo>'}",
            f"  title:    {title}",
            f"  labels:   {' '.join(_labels(ticket))}",
            f"  assignee: {assignee or '<none>'}",
            "--- body ---",
            body,
        ]
        return {"preview": "\n".join(preview)}

    dup = next((i for i in _list_issues(title, repo) if i.get("title") == title), None)
    if dup:
        sys.stderr.write(
            f"github: an open issue with this title already exists: {dup['url']} (skipping create)\n")
        return {"provider": "github", "ref": str(dup["number"]), "url": dup["url"], "deduped": True}

    def create(with_labels: bool) -> subprocess.CompletedProcess:
        cmd = ["gh", "issue", "create"]
        if repo:
            cmd += ["--repo", repo]
        cmd += ["--title", title, "--body-file", "-"]
        if with_labels:
            for l in labels:
                cmd += ["--label", l]
        if assignee:
            cmd += ["--assignee", assignee]
        return subprocess.run(cmd, input=body, capture_output=True, text=True)

    r = create(bool(labels))
    if r.returncode != 0 and labels and "label" in (r.stderr or "").lower():
        sys.stderr.write(
            "github: some labels do not exist in the repo; retrying without labels "
            "(type/priority stay in the body)\n")
        r = create(False)
    if r.returncode != 0:
        raise RuntimeError((r.stderr or "").strip() or "gh issue create failed")

    url = r.stdout.strip().splitlines()[-1].strip()
    m = re.search(r"(\d+)\s*$", url)
    return {"provider": "github", "ref": m.group(1) if m else "", "url": url}
